### PR TITLE
Implement SDL-0221 Allow Multiple Modules per Module Type

### PR DESCRIPTION
**Pull request by Tnnnguyen**
Fixes [Proposal SDL-0221](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0221-multiple-modules.md)

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit tests, test with new data from Core

### Summary
Parse new Module Control data, store and send to app when requested

### Tasks Remaining:
- [x] Add more unit tests

### CLA
- [ x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by Tnnnguyen** [comment]
@theresalech this PR is ready for review.  I believe the failed test suite is due to server setup and not the tests themselves.

**Comment by theresalech** [comment]
Fixes issue #1070

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package com.smartdevicelink.proxy.rpc;
+
+import android.security.KeyChain;
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java]: This import is unused and probably is what is causing the JavaSE tests to fail

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package com.smartdevicelink.test.rpc.datatypes;
+
+import com.smartdevicelink.marshal.JsonRPCMarshaller;
+import com.smartdevicelink.proxy.rpc.Grid;
+import com.smartdevicelink.proxy.rpc.ModuleInfo;
+import com.smartdevicelink.test.JsonUtils;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java]: unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import com.smartdevicelink.test.Test;
+import com.smartdevicelink.test.Validator;
+
+import junit.framework.TestCase;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Hashtable;
+import java.util.Iterator;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java]: unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/SeatLocationCapabilityTests.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
+package com.smartdevicelink.test.rpc.datatypes;
+
+import android.util.Log;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/SeatLocationCapabilityTests.java]: multiple unused imports

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/requests/GetInteriorVehicleDataConsentTests.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import com.smartdevicelink.protocol.enums.FunctionID;
+import com.smartdevicelink.proxy.RPCMessage;
+import com.smartdevicelink.proxy.rpc.GetInteriorVehicleDataConsent;
+import com.smartdevicelink.proxy.rpc.enums.ModuleType;
+import com.smartdevicelink.test.BaseRpcTests;
+import com.smartdevicelink.test.JsonUtils;
+import com.smartdevicelink.test.Test;
+import com.smartdevicelink.test.json.rpc.JsonFileReader;
+
+import junit.framework.TestCase;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/requests/GetInteriorVehicleDataConsentTests.java]: unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetInteriorVehicleDataConsentResponseTests.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
+package com.smartdevicelink.test.rpc.responses;
+
+import com.smartdevicelink.marshal.JsonRPCMarshaller;
+import com.smartdevicelink.protocol.enums.FunctionID;
+import com.smartdevicelink.proxy.RPCMessage;
+import com.smartdevicelink.proxy.rpc.GetInteriorVehicleDataConsent;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetInteriorVehicleDataConsentResponseTests.java]: unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/ReleaseInteriorVehicleDataModuleResponseTests.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package com.smartdevicelink.test.rpc.responses;
+
+import com.smartdevicelink.marshal.JsonRPCMarshaller;
+import com.smartdevicelink.protocol.enums.FunctionID;
+import com.smartdevicelink.proxy.RPCMessage;
+import com.smartdevicelink.proxy.rpc.GetInteriorVehicleDataConsentResponse;
```
Comment (BrettyWhite) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/ReleaseInteriorVehicleDataModuleResponseTests.java]: unused import

File: base/src/main/java/com/smartdevicelink/proxy/rpc/ButtonCapabilities.java [hunk 8]
```diff
@@ -97,6 +97,7 @@
 	public static final String KEY_SHORT_PRESS_AVAILABLE = "shortPressAvailable";
 	public static final String KEY_LONG_PRESS_AVAILABLE = "longPressAvailable";
 	public static final String KEY_UP_DOWN_AVAILABLE = "upDownAvailable";
+    public static final String KEY_MODULE_INFO = "moduleInfo";
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/ButtonCapabilities.java]: indentation

File: base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsent.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public void setModuleIds(List<String> ids) {
+        setParameters(KEY_MODULE_ID, ids);
+    }
+
+    /**
+     * Gets the Module Ids of this class
+     * @return the Module Ids
+     */
+    public List<String> getModuleIds() {
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsent.java]: We can suppress the warning here with `@SuppressWarnings("unchecked")`

File: base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsentResponse.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public void setAllowances(List<Boolean> allowances) {
+        setParameters(KEY_ALLOWED, allowances);
+    }
+
+    /**
+     * Gets the list of allowances of this class
+     * @return the list of allowances of this class
+     */
+    public List<Boolean> getAllowances() {
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsentResponse.java]: We can suppress the warning here with `@SuppressWarnings("unchecked")`

File: base/src/main/java/com/smartdevicelink/proxy/rpc/ReleaseInteriorVehicleDataModule.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
+package com.smartdevicelink.proxy.rpc;
+
+import com.smartdevicelink.protocol.enums.FunctionID;
+import com.smartdevicelink.proxy.RPCRequest;
+import com.smartdevicelink.proxy.rpc.enums.ModuleType;
+
+import java.util.Hashtable;
+import java.util.List;
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/ReleaseInteriorVehicleDataModule.java]: unused import

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    public void setSeats(List<SeatLocation> locations) {
+        setValue(KEY_SEATS, locations);
+    }
+
+    /**
+     * Gets the seat locations of this capability
+     * @return the seat locations
+     */
+    public List<SeatLocation> getSeatLocations() {
```
Comment (BrettyWhite) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java]: We can suppress the warning here with `@SuppressWarnings("unchecked")`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/Test.java [hunk 13]
```diff
@@ -75,6 +78,8 @@
 import com.smartdevicelink.proxy.rpc.SdlMsgVersion;
 import com.smartdevicelink.proxy.rpc.SeatControlCapabilities;
 import com.smartdevicelink.proxy.rpc.SeatControlData;
+import com.smartdevicelink.proxy.rpc.SeatLocation;
+import com.smartdevicelink.proxy.rpc.SeatLocationCapability;
```
[the remark this answers is not in the document]
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/Test.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/Validator.java [hunk 14]
```diff
@@ -72,6 +74,8 @@
 import com.smartdevicelink.proxy.rpc.SdlMsgVersion;
 import com.smartdevicelink.proxy.rpc.SeatControlCapabilities;
 import com.smartdevicelink.proxy.rpc.SeatControlData;
+import com.smartdevicelink.proxy.rpc.SeatLocation;
+import com.smartdevicelink.proxy.rpc.SeatLocationCapability;
```
[the remark this answers is not in the document]
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/Validator.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java (hunk 2 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java (hunk 3 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/ModuleInfoTests.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/SeatLocationCapabilityTests.java (hunk 4 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/SeatLocationCapabilityTests.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/requests/GetInteriorVehicleDataConsentTests.java (hunk 5 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/requests/GetInteriorVehicleDataConsentTests.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/ReleaseInteriorVehicleDataModuleResponseTests.java (hunk 7 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/ReleaseInteriorVehicleDataModuleResponseTests.java]: Done

File: base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsentResponse.java (hunk 10 above)
Comment (Tnnnguyen) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsentResponse.java]: Done

File: base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsent.java (hunk 9 above)
Comment (Tnnnguyen) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/GetInteriorVehicleDataConsent.java]: Done

File: base/src/main/java/com/smartdevicelink/proxy/rpc/ReleaseInteriorVehicleDataModule.java (hunk 11 above)
Comment (Tnnnguyen) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/ReleaseInteriorVehicleDataModule.java]: Done

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java (hunk 12 above)
Comment (Tnnnguyen) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/SeatLocationCapability.java]: Done

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetInteriorVehicleDataConsentResponseTests.java (hunk 6 above)
Comment (Tnnnguyen) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetInteriorVehicleDataConsentResponseTests.java]: Done

**Comment by bilal-alsharifi** [comment]
@Tnnnguyen Can you please fix the merge conflict before we merge the PR.

**Comment by theresalech** [comment]
@Tnnnguyen please follow the steps Brett provided to properly reset and merge develop into this branch.

**Comment by Tnnnguyen** [comment]
> @Tnnnguyen please follow the steps Brett provided to properly reset and merge develop into this branch.

@theresalech will do

**Comment by BrettyWhite** [comment]
Implemented in #1166

**Pull request closed without merging**